File: src/lib/api/ttsClient.ts
/**
 * HTTP client for the TTS sidecar server.
 */

const BASE_URL = "http://127.0.0.1:8765";

export interface HealthResponse {
  status: string;
  version: string;
}

export interface ModelStatus {
  loaded: boolean;
  model_id: string | null;
  device: string | null;
  memory: {
    device: string;
    total_gb: number;
    available_gb: number;
  };
}

export interface StartupStatus {
  phase: string;
  message: string;
  progress: number;
}

export interface DownloadProgress {
  status: "idle" | "downloading" | "complete" | "error";
  file_name: string;
  bytes_downloaded: number;
  bytes_total: number;
  speed_mbps: number;
  eta: number;
}

export interface SystemInfo {
  python_version: string;
  torch_version: string;
  device: string;
  device_name: string;
  memory_total_gb: number;
  memory_available_gb: number;
  cache_dir: string;
}

export interface LogEntry {
  level: string;
  message: string;
  timestamp: string;
}

export interface SpeakerInfo {
  name: string;
  description: string;
  native_language: string;
  personality: string;
  gender: string;
}

export interface CustomVoiceRequest {
  text: string;
  speaker: string;
  instruction?: string;
  language?: string;
  format?: string;
  stable_lead_in?: boolean;
  seed?: number | null;
  sample_rate?: number | null;
  bit_depth?: number;
}

export interface VoiceDesignRequest {
  text: string;
  voice_description: string;
  language?: string;
  format?: string;
  stable_lead_in?: boolean;
  seed?: number | null;
  sample_rate?: number | null;
  bit_depth?: number;
}

export interface VoiceCloneRequestOptions {
  xVectorOnly?: boolean;
  language?: string;
  format?: string;
  seed?: number | null;
  sample_rate?: number | null;
  bit_depth?: number;
}

export interface BatchItem {
  text: string;
  output_filename: string;
}

export interface BatchRequest {
  mode: "custom-voice" | "voice-clone" | "voice-design";
  language?: string;
  format?: string;
  mp3_bitrate?: number;
  seed?: number | null;
  sample_rate?: number | null;
  bit_depth?: number;
  speaker?: string;
  instruction?: string;
  voice_description?: string;
  stable_lead_in?: boolean;
  reference_text?: string;
  reference_audio_base64?: string | null;
  x_vector_only_mode?: boolean;
  items: BatchItem[];
}

export interface BatchProgress {
  total: number;
  completed: number;
  current_item: string;
  status: string;
}

export interface WhisperStatus {
  loaded: boolean;
  model_size: string | null;
  device: string;
}

export interface WhisperModelInfo {
  size: string;
  parameters: string;
  download_size_mb: number;
}

export interface TranslationStatus {
  loaded: boolean;
  model_key: string | null;
  model_id: string | null;
  device: string;
}

export interface TranslationModelInfo {
  key: string;
  label: string;
  model_id: string;
  parameters: string;
  download_size_mb: number;
}

export interface TranscriptionResult {
  text: string;
  language: string;
  confidence: number;
  duration_seconds: number;
}

export type TranscriptionTask = "transcribe" | "translate";

export interface TranslateTextResult {
  text: string;
  source_language: string;
  target_language: string;
}

export const PRESET_SPEAKERS = [
  "aiden",
  "dylan",
  "eric",
  "ono_anna",
  "ryan",
  "serena",
  "sohee",
  "uncle_fu",
  "vivian",
] as const;

export type Speaker = (typeof PRESET_SPEAKERS)[number];

const GENERATION_TIMEOUT_MS = 5 * 60 * 1000; // 5 minutes

class TTSClient {
  private baseUrl: string;
  private authToken: string | null = null;
  private _abortController: AbortController | null = null;

  constructor(baseUrl: string = BASE_URL) {
    this.baseUrl = baseUrl;
  }

  setAccessToken(token: string | null): void {
    const trimmed = token?.trim();
    this.authToken = trimmed ? trimmed : null;
  }

  private async request(path: string, init: RequestInit = {}): Promise<Response> {
    if (!this.authToken) {
      if (Object.keys(init).length === 0) {
        return fetch(`${this.baseUrl}${path}`);
      }
      return fetch(`${this.baseUrl}${path}`, init);
    }

    const headers = new Headers(init.headers);
    headers.set("X-API-Key", this.authToken);
    return fetch(`${this.baseUrl}${path}`, { ...init, headers });
  }

  /** Create a signal for generation requests with timeout + manual abort. */
  private createGenerationSignal(): AbortSignal {
    this._abortController = new AbortController();
    const timeoutId = setTimeout(() => this._abortController?.abort("Generation timed out"), GENERATION_TIMEOUT_MS);
    // Clear timeout when signal aborts (manual or timeout)
    this._abortController.signal.addEventListener("abort", () => clearTimeout(timeoutId), { once: true });
    return this._abortController.signal;
  }

  /** Abort the current generation request. */
  abortGeneration(): void {
    this._abortController?.abort("Generation cancelled");
    this._abortController = null;
  }

  get isGenerating(): boolean {
    return this._abortController !== null;
  }

  private async readErrorMessage(res: Response, fallback: string): Promise<string> {
    try {
      const data = await res.json();
      const detail = (data as { detail?: unknown })?.detail;
      if (typeof detail === "string") return detail;
      if (Array.isArray(detail)) {
        const messages = detail
          .map((item) => {
            if (!item) return null;
            if (typeof item === "string") return item;
            if (typeof item === "object" && "msg" in item) {
              const msg = (item as { msg?: string }).msg;
              return msg ?? null;
            }
            return null;
          })
          .filter(Boolean);
        if (messages.length) return messages.join("; ");
      }
      if (detail && typeof detail === "object") {
        const obj = detail as Record<string, unknown>;
        if (typeof obj.message === "string") return obj.message;
        if (typeof obj.msg === "string") return obj.msg;
        return JSON.stringify(detail);
      }
      if (detail) return String(detail);
      return fallback;
    } catch {
      return fallback;
    }
  }

  // ============================================================================
  // Health & Status
  // ============================================================================

  async health(): Promise<HealthResponse> {
    const res = await this.request("/health");
    if (!res.ok) throw new Error("Server not available");
    return res.json();
  }

  async getStartupStatus(): Promise<StartupStatus> {
    const res = await this.request("/startup-status");
    if (!res.ok) {
      const error = new Error("Failed to get startup status") as Error & { status?: number };
      error.status = res.status;
      throw error;
    }
    return res.json();
  }

  async getDownloadProgress(): Promise<DownloadProgress> {
    const res = await this.request("/download-progress");
    if (!res.ok) throw new Error("Failed to get download progress");
    return res.json();
  }

  async getModelStatus(): Promise<ModelStatus> {
    const res = await this.request("/model-status");
    if (!res.ok) throw new Error("Failed to get model status");
    return res.json();
  }

  // ============================================================================
  // Debug & System Info
  // ============================================================================

  async getSystemInfo(): Promise<SystemInfo> {
    const res = await this.request("/system-info");
    if (!res.ok) throw new Error("Failed to get system info");
    return res.json();
  }

  async getLogs(count: number = 100, level?: string): Promise<LogEntry[]> {
    const params = new URLSearchParams({ count: count.toString() });
    if (level) params.append("level", level);

    const res = await this.request(`/logs?${params}`);
    if (!res.ok) throw new Error("Failed to get logs");
    return res.json();
  }

  // ============================================================================
  // Speaker & Voice Info
  // ============================================================================

  async getSpeakers(): Promise<string[]> {
    const res = await this.request("/speakers");
    if (!res.ok) throw new Error("Failed to get speakers");
    const data = await res.json();
    return data.speakers;
  }

  async getSpeakersInfo(): Promise<SpeakerInfo[]> {
    const res = await this.request("/speakers-info");
    if (!res.ok) throw new Error("Failed to get speaker info");
    return res.json();
  }

  async getLanguages(): Promise<string[]> {
    const res = await this.request("/languages");
    if (!res.ok) throw new Error("Failed to get languages");
    const data = await res.json();
    return data.languages;
  }

  // ============================================================================
  // Model Management
  // ============================================================================

  async loadModel(modelId: string = "0.6b"): Promise<void> {
    const res = await this.request("/load-model", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ model_id: modelId }),
    });
    if (!res.ok) {
      throw new Error(await this.readErrorMessage(res, "Failed to load model"));
    }
  }

  async unloadModel(): Promise<void> {
    const res = await this.request("/unload-model", {
      method: "POST",
    });
    if (!res.ok) throw new Error("Failed to unload model");
  }

  // ============================================================================
  // Generation
  // ============================================================================

  async generateCustomVoice(request: CustomVoiceRequest): Promise<Blob> {
    const signal = this.createGenerationSignal();
    try {
      const res = await this.request("/generate/custom-voice", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(request),
        signal,
      });
      if (!res.ok) {
        throw new Error(await this.readErrorMessage(res, "Failed to generate audio"));
      }
      return res.blob();
    } finally {
      this._abortController = null;
    }
  }

  async generateVoiceClone(
    text: string,
    referenceText: string,
    referenceAudio: File,
    options?: VoiceCloneRequestOptions
  ): Promise<Blob> {
    const signal = this.createGenerationSignal();
    const formData = new FormData();
    formData.append("text", text);
    formData.append("reference_text", referenceText ?? "");
    formData.append("reference_audio", referenceAudio);
    formData.append("x_vector_only_mode", options?.xVectorOnly ? "true" : "false");
    if (options?.language) formData.append("language", options.language);
    if (options?.format) formData.append("format", options.format);
    if (typeof options?.seed === "number") formData.append("seed", options.seed.toString());
    if (typeof options?.sample_rate === "number") {
      formData.append("sample_rate", options.sample_rate.toString());
    }
    if (typeof options?.bit_depth === "number") {
      formData.append("bit_depth", options.bit_depth.toString());
    }

    try {
      const res = await this.request("/generate/voice-clone", {
        method: "POST",
        body: formData,
        signal,
      });
      if (!res.ok) {
        throw new Error(await this.readErrorMessage(res, "Failed to generate audio"));
      }
      return res.blob();
    } finally {
      this._abortController = null;
    }
  }

  async generateVoiceDesign(request: VoiceDesignRequest): Promise<Blob> {
    const signal = this.createGenerationSignal();
    try {
      const res = await this.request("/generate/voice-design", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(request),
        signal,
      });
      if (!res.ok) {
        throw new Error(await this.readErrorMessage(res, "Failed to generate audio"));
      }
      return res.blob();
    } finally {
      this._abortController = null;
    }
  }

  async generateBatch(request: BatchRequest): Promise<Blob> {
    const signal = this.createGenerationSignal();
    try {
      const res = await this.request("/generate/batch", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(request),
        signal,
      });
      if (!res.ok) {
        throw new Error(await this.readErrorMessage(res, "Failed to generate batch"));
      }
      return res.blob();
    } finally {
      this._abortController = null;
    }
  }

  async getBatchProgress(): Promise<BatchProgress> {
    const res = await this.request("/batch-progress");
    if (!res.ok) throw new Error("Failed to get batch progress");
    return res.json();
  }

  async cancelBatch(): Promise<void> {
    const res = await this.request("/cancel-generation", { method: "POST" });
    if (!res.ok) throw new Error("Failed to cancel batch");
  }

  // ============================================================================
  // Whisper Transcription
  // ============================================================================

  async whisperStatus(): Promise<WhisperStatus> {
    const res = await this.request("/whisper-status");
    if (!res.ok) throw new Error("Failed to get Whisper status");
    return res.json();
  }

  async whisperModels(): Promise<WhisperModelInfo[]> {
    const res = await this.request("/whisper-models");
    if (!res.ok) throw new Error("Failed to get Whisper models");
    return res.json();
  }

  async loadWhisper(modelSize: string = "base"): Promise<void> {
    const res = await this.request("/load-whisper", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ model_size: modelSize }),
    });
    if (!res.ok) {
      throw new Error(await this.readErrorMessage(res, "Failed to load Whisper model"));
    }
  }

  async unloadWhisper(): Promise<void> {
    const res = await this.request("/unload-whisper", {
      method: "POST",
    });
    if (!res.ok) throw new Error("Failed to unload Whisper model");
  }

  async translationStatus(): Promise<TranslationStatus> {
    const res = await this.request("/translation-status");
    if (!res.ok) throw new Error("Failed to get translation status");
    return res.json();
  }

  async translationModels(): Promise<TranslationModelInfo[]> {
    const res = await this.request("/translation-models");
    if (!res.ok) throw new Error("Failed to get translation models");
    return res.json();
  }

  async loadTranslation(modelKey: string = "nllb-600m"): Promise<void> {
    const res = await this.request("/load-translation", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ model_key: modelKey }),
    });
    if (!res.ok) {
      throw new Error(await this.readErrorMessage(res, "Failed to load translation model"));
    }
  }

  async unloadTranslation(): Promise<void> {
    const res = await this.request("/unload-translation", {
      method: "POST",
    });
    if (!res.ok) throw new Error("Failed to unload translation model");
  }

  async transcribe(
    audioFile: File | Blob,
    options?: { task?: TranscriptionTask },
  ): Promise<TranscriptionResult> {
    const formData = new FormData();
    const file = audioFile instanceof File
      ? audioFile
      : new File([audioFile], "audio.wav", { type: audioFile.type || "audio/wav" });
    formData.append("audio", file);
    if (options?.task) {
      formData.append("task", options.task);
    }

    const res = await this.request("/transcribe", {
      method: "POST",
      body: formData,
    });
    if (!res.ok) {
      throw new Error(await this.readErrorMessage(res, "Failed to transcribe audio"));
    }
    return res.json();
  }

  async translateText(
    text: string,
    targetLanguage: string,
    sourceLanguage: string = "auto",
  ): Promise<TranslateTextResult> {
    const res = await this.request("/translate-text", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        text,
        target_language: targetLanguage,
        source_language: sourceLanguage,
      }),
    });
    if (!res.ok) {
      throw new Error(await this.readErrorMessage(res, "Failed to translate text"));
    }
    return res.json();
  }

  // ============================================================================
  // Lifecycle
  // ============================================================================

  async shutdown(): Promise<void> {
    await this.request("/shutdown", { method: "POST" });
  }
}

export const ttsClient = new TTSClient();

File: privatevoice-control/mcp-server/privatevoice_mcp/tts_client.py
"""HTTP client for the PrivateVoice FastAPI backend (port 8765)."""

from __future__ import annotations

import base64
import os
import tempfile

import httpx


class TTSClient:
    """Wraps all FastAPI backend endpoints."""

    def __init__(self, base_url: str, timeout: float = 120.0) -> None:
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout

    def _client(self) -> httpx.AsyncClient:
        return httpx.AsyncClient(base_url=self.base_url, timeout=self.timeout)

    # ------------------------------------------------------------------
    # Health / status
    # ------------------------------------------------------------------

    async def check_health(self) -> dict:
        try:
            async with self._client() as c:
                r = await c.get("/health", timeout=5.0)
                data = r.json()
                data["alive"] = True
                return data
        except (httpx.ConnectError, httpx.TimeoutException, httpx.HTTPError):
            return {"alive": False}

    async def get_startup_status(self) -> dict:
        async with self._client() as c:
            r = await c.get("/startup-status")
            return r.json()

    async def get_model_status(self) -> dict:
        async with self._client() as c:
            r = await c.get("/model-status")
            return r.json()

    async def get_model_catalog(self) -> dict:
        async with self._client() as c:
            r = await c.get("/model-catalog")
            return r.json()

    async def get_system_info(self) -> dict:
        async with self._client() as c:
            r = await c.get("/system-info")
            return r.json()

    async def get_composite_status(self) -> dict:
        """Fetch health, startup-status, model-status, system-info in parallel."""
        async with self._client() as c:
            results = {}
            for name, path in [
                ("health", "/health"),
                ("startup", "/startup-status"),
                ("model", "/model-status"),
                ("system", "/system-info"),
            ]:
                try:
                    r = await c.get(path, timeout=10.0)
                    results[name] = r.json()
                except Exception as exc:
                    results[name] = {"error": str(exc)}
            return results

    # ------------------------------------------------------------------
    # Model management
    # ------------------------------------------------------------------

    async def load_model(
        self,
        model_id: str | None = None,
        provider: str | None = None,
        model_key: str | None = None,
    ) -> dict:
        payload: dict[str, str] = {}
        if model_id:
            payload["model_id"] = model_id
        if provider:
            payload["provider"] = provider
        if model_key:
            payload["model_key"] = model_key

        if not payload:
            return {"error": "Provide model_id or provider/model_key."}

        async with self._client() as c:
            r = await c.post("/load-model", json=payload)
            return r.json()

    async def unload_model(self) -> dict:
        async with self._client() as c:
            r = await c.post("/unload-model")
            return r.json()

    # ------------------------------------------------------------------
    # Speakers
    # ------------------------------------------------------------------

    async def get_speakers(self) -> dict:
        async with self._client() as c:
            r = await c.get("/speakers")
            return r.json()

    # ------------------------------------------------------------------
    # Generation
    # ------------------------------------------------------------------

    async def generate_custom_voice(self, params: dict) -> dict:
        text = params.get("text")
        if not text:
            return {"error": "Parameter 'text' is required."}

        payload = {
            "text": text,
            "speaker": params.get("speaker", "serena"),
            "instruction": params.get("instruction", ""),
            "language": params.get("language", "english"),
            "format": params.get("format", "wav"),
            "mp3_bitrate": params.get("mp3_bitrate", 192),
        }

        return await self._generate("/generate/custom-voice", payload)

    async def generate_voice_clone(self, params: dict) -> dict:
        text = params.get("text")
        ref_audio_path = params.get("reference_audio")
        if not text:
            return {"error": "Parameter 'text' is required."}
        if not ref_audio_path:
            return {"error": "Parameter 'reference_audio' is required."}
        if not os.path.exists(ref_audio_path):
            return {"error": f"Reference audio not found: {ref_audio_path}"}

        form_data = {
            "text": text,
            "reference_text": params.get("reference_text", ""),
            "x_vector_only_mode": str(params.get("x_vector_only_mode", False)).lower(),
            "language": params.get("language", "english"),
            "format": params.get("format", "wav"),
            "mp3_bitrate": str(params.get("mp3_bitrate", 192)),
        }

        async with self._client() as c:
            with open(ref_audio_path, "rb") as f:
                files = {"reference_audio": (os.path.basename(ref_audio_path), f, "audio/wav")}
                r = await c.post("/generate/voice-clone", data=form_data, files=files)

            if r.status_code != 200:
                return {"error": f"Generation failed ({r.status_code}): {r.text}"}

            return self._save_audio(r, params.get("format", "wav"))

    async def generate_voice_design(self, params: dict) -> dict:
        text = params.get("text")
        voice_desc = params.get("voice_description")
        if not text:
            return {"error": "Parameter 'text' is required."}
        if not voice_desc:
            return {"error": "Parameter 'voice_description' is required."}

        payload = {
            "text": text,
            "voice_description": voice_desc,
            "language": params.get("language", "english"),
            "format": params.get("format", "wav"),
            "mp3_bitrate": params.get("mp3_bitrate", 192),
        }

        return await self._generate("/generate/voice-design", payload)

    async def generate_speech(self, params: dict) -> dict:
        mode = params.get("mode")
        text = params.get("text")
        if not mode:
            return {"error": "Parameter 'mode' is required."}
        if not text:
            return {"error": "Parameter 'text' is required."}

        payload = {
            "mode": mode,
            "provider": params.get("provider"),
            "model_key": params.get("model_key"),
            "text": text,
            "language": params.get("language", "english"),
            "speaker": params.get("speaker", "serena"),
            "instruction": params.get("instruction", ""),
            "voice_description": params.get("voice_description", ""),
            "reference_text": params.get("reference_text", ""),
            "x_vector_only_mode": params.get("x_vector_only_mode", False),
            "format": params.get("format", "wav"),
            "mp3_bitrate": params.get("mp3_bitrate", 192),
            "stable_lead_in": params.get("stable_lead_in", True),
            "seed": params.get("seed"),
            "sample_rate": params.get("sample_rate"),
            "bit_depth": params.get("bit_depth", 16),
            "advanced": params.get("advanced", {}) or {},
            "reference_audio_base64": None,
        }

        ref_audio_path = params.get("reference_audio")
        if ref_audio_path:
            if not os.path.exists(ref_audio_path):
                return {"error": f"Reference audio not found: {ref_audio_path}"}
            with open(ref_audio_path, "rb") as f:
                payload["reference_audio_base64"] = base64.b64encode(f.read()).decode("ascii")

        return await self._generate("/generate/speech", payload)

    # ------------------------------------------------------------------
    # Shutdown
    # ------------------------------------------------------------------

    async def shutdown_backend(self) -> dict:
        try:
            async with self._client() as c:
                r = await c.post("/shutdown", timeout=10.0)
                return r.json()
        except Exception as exc:
            return {"error": str(exc)}

    # ------------------------------------------------------------------
    # Action dispatcher
    # ------------------------------------------------------------------

    async def trigger_action(self, action: str, params: dict) -> dict:
        dispatch = {
            "generate_custom_voice": self.generate_custom_voice,
            "generate_voice_clone": self.generate_voice_clone,
            "generate_voice_design": self.generate_voice_design,
            "generate_speech": self.generate_speech,
            "load_model": lambda p: self.load_model(
                model_id=p.get("model_id"),
                provider=p.get("provider"),
                model_key=p.get("model_key"),
            ),
            "load_provider_model": lambda p: self.load_model(
                provider=p.get("provider"),
                model_key=p.get("model_key"),
                model_id=p.get("model_id"),
            ),
            "unload_model": lambda _: self.unload_model(),
            "get_speakers": lambda _: self.get_speakers(),
            "get_model_status": lambda _: self.get_model_status(),
            "get_model_catalog": lambda _: self.get_model_catalog(),
            "get_system_info": lambda _: self.get_system_info(),
        }

        handler = dispatch.get(action)
        if handler is None:
            return {"error": f"Unknown action: {action}"}

        try:
            return await handler(params)
        except httpx.ConnectError:
            return {
                "error": "Backend unreachable. Is PrivateVoice running? Use app_launch_and_verify first."
            }
        except Exception as exc:
            return {"error": f"Action '{action}' failed: {exc}"}

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    async def _generate(self, endpoint: str, payload: dict) -> dict:
        """POST JSON to a generation endpoint, save the returned audio."""
        async with self._client() as c:
            r = await c.post(endpoint, json=payload)

            if r.status_code != 200:
                return {"error": f"Generation failed ({r.status_code}): {r.text}"}

            return self._save_audio(r, payload.get("format", "wav"))

    @staticmethod
    def _save_audio(response: httpx.Response, fmt: str = "wav") -> dict:
        ext = fmt if fmt in ("wav", "mp3") else "wav"
        fd, path = tempfile.mkstemp(suffix=f".{ext}", prefix="pv_generated_")
        os.close(fd)

        with open(path, "wb") as f:
            f.write(response.content)

        return {
            "audio_path": path,
            "format": ext,
            "size_bytes": len(response.content),
            "content_type": response.headers.get("content-type", "unknown"),
        }
